src/v2/dry_run.rs

```diff
@@ -0,0 +1,1063 @@
+use concordium_base::{
+    base::{Energy, ProtocolVersion},
+    common::types::{CredentialIndex, KeyIndex, Timestamp},
+    contracts_common::{AccountAddress, Amount, ContractAddress},
+    hashes::BlockHash,
+    smart_contracts::ContractTraceElement,
+    transactions::{EncodedPayload, PayloadLike},
+};
+
+use futures::*;
+use tonic::metadata::MetadataValue;
+
+use crate::{
+    types::{
+        smart_contracts::{ContractContext, InstanceInfo, ReturnValue},
+        AccountInfo, AccountTransactionDetails, RejectReason,
+    },
+    v2::{generated::DryRunStateQuery, Require},
+};
+
+use super::{
+    generated::{
+        self, account_transaction_payload, dry_run_error_response, dry_run_request,
+        DryRunInvokeInstance, DryRunMintToAccount, DryRunSignature, DryRunStateOperation,
+    },
+    AccountIdentifier, IntoBlockIdentifier,
+};
+
+mod shared_receiver {
```

Comment:
This seems to be unnecessarily complex for what it achieves. Especially the use of locking seems less than ideal, and the API of this DryRun is rather complex and unergonomic for common uses.

What I'd recommend is to use tasks and message passing more rather than locking.

When you construct a `DryRun` client you ask for the block state since that is necessary to do anything useful with it.

Then in that method you spawn two new tasks (and record its handle) whose job is to

- task 1 sends listens to a channel for (tokio::oneshot::Sender<>, Request) values and when it receives them it sends the request to the node (as in, enqueues it to the stream), and sends the sender through channel B
- task 2 listens on the incoming request stream and the channel B together (e.g., using `zip` on streams) and when it receives a value it sends it to the Sender.

When issuing a request you just return a `tokio::oneshot::Receiver` part that the user can listen on. That'll simplify the setup and make it in-line with the message passing style we use elsewhere.

The two tasks should be cancelled in the Drop implementation of DryRun.

Additionally, to make the API of DryRun more ergonomic the default methods should be blocking in the sense that when I do 
```
dry_run.load_block_state().await?
```
I get the response immediately.

We can have additional "send_load_block_state` methods (or some other name if you have a better idea) that return the future similar to what they do now.